pkg/archive: sort files

vbatts opened the pull request:
This will sort changes found, and the files found for Tar() ( and
TarWithOptions()).

Sorting the files before appending them to the tar archive
would mean a dependable ordering for types like hardlinks.

See also https://gist.github.com/vbatts/d35e9f18fbfac74b36f4

Signed-off-by: Vincent Batts vbatts@redhat.com

vbatts committed:
pkg/archive: sort files

sort changes found and exported.

Sorting the files before appending them to the tar archive
would mean a dependable ordering for types like hardlinks.

Also, combine sort logic used

Signed-off-by: Vincent Batts <vbatts@redhat.com>

vbatts commented:
ping @unclejack @tiborvass @dmcgowan @jlhawn

unclejack commented:
@vbatts Can you add some tests for this, please? Regressing on this in the future would be really bad.

vbatts committed:
pkg/archive: test that confirms hardlink ordering

This test was written against
master(abdfb21e3a761efdd70614de42905ff7911c5372) as a failing test, but
works with this patch set.

Signed-off-by: Vincent Batts <vbatts@redhat.com>

vbatts commented:
added a test that fails on master, but this patch fixes it.

vbatts commented:
interesting side effect is that the layer's side checksum is more deterministic ... ;-)

```
vbatts@valse ~ (master *) $ /home/vbatts/src/docker/docker/bundles/1.4.1-dev/dynbinary/docker-1.4.1-dev save ubuntu:utopic | tar Ox ef3ba9f35b97ff1ba5c14a82ce512cf9df8736a3bed3ca44ea3a35dddc3c1019/layer.tar | md5sum -
200a5aa092a580dc528d6bed1152b384  -
vbatts@valse ~ (master *) $ /home/vbatts/src/docker/docker/bundles/1.4.1-dev/dynbinary/docker-1.4.1-dev save ubuntu:utopic | tar Ox ef3ba9f35b97ff1ba5c14a82ce512cf9df8736a3bed3ca44ea3a35dddc3c1019/layer.tar | md5sum -
200a5aa092a580dc528d6bed1152b384  -
vbatts@valse ~ (master *) $ /home/vbatts/src/docker/docker/bundles/1.4.1-dev/dynbinary/docker-1.4.1-dev save ubuntu:utopic | tar Ox ef3ba9f35b97ff1ba5c14a82ce512cf9df8736a3bed3ca44ea3a35dddc3c1019/layer.tar | md5sum -
200a5aa092a580dc528d6bed1152b384  -
vbatts@valse ~ (master *) $ /home/vbatts/src/docker/docker/bundles/1.4.1-dev/dynbinary/docker-1.4.1-dev save ubuntu:utopic | tar Ox ef3ba9f35b97ff1ba5c14a82ce512cf9df8736a3bed3ca44ea3a35dddc3c1019/layer.tar | md5sum -
200a5aa092a580dc528d6bed1152b384  -
```

tiborvass commented on pkg/archive/archive.go:
```diff
@@ -457,13 +460,30 @@ func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error)
 
+		// sort dirsToAdd and pathsToAdd
+		sort.Sort(PathsToAddByRelPath(dirsToAdd))
+		sort.Sort(PathsToAddByRelPath(pathsToAdd))
```
I wonder if we could/should make this concurrent

vbatts replied:
could, sure. It's not _that_ expensive though.

tiborvass replied:
ok then let's leave it out for now

vbatts commented:
@tiborvass the lack of determinism is in `Changes()` and consequently what `ExportChanges()` is providing. I've removed the sorting around `filepath.Walk()`.

LK4D4 commented on pkg/archive/changes_posix_test.go:
```diff
@@ -0,0 +1,126 @@
+func (th TarHeaders) Swap(i, j int)      { th[j], th[i] = th[i], th[j] }
+func (th TarHeaders) Less(i, j int) bool { return th[i].Name < th[j].Name }
+
+func WalkHeaders(r io.Reader) (headers []tar.Header, err error) {
```
is here really necessary to use named returns? I don't like them because you need always remember scoping rules :)
Also shouldn't this be private?

LK4D4 commented on pkg/archive/changes_posix_test.go:
```diff
@@ -0,0 +1,126 @@
+
+}
+
+type TarHeaders []tar.Header
```
Maybe all this types should be private?

vbatts commented:
@LK4D4 fixed. PTAL

jlhawn commented:
@vbatts I know it's been removed now, but an earlier version of your patch sorted and separated dirs and non-dirs in TarWithOptions. This was unnecessary because of it's use of `filepath.Walk()`, but it got me thinking: we *should* at least sort `options.IncludeFiles`. What do you think?

dmcgowan commented:
Confirmed this fixes my push issue with ubuntu:topic to a v2 registry

crosbymichael commented:
LGTM

jfrazelle commented:
I can confirm that the tests fail on master LGTM

tiborvass commented:
LGTM
